### ISSUE TITLE
Audit : Handle file permissions

### ISSUE THREAD

**Issue by V-ed** [Vhoyon/Discord-Bot#128]
After the PR #127 has been merged, a new audit module is now in our bot, but assumes that the file is writable (which is a bad assumption) and that the file is its audit file. Currently, the only handling is through a try-catch block (in [this file](https://github.com/Vhoyon/Discord-Bot/blob/dev/framework/vendor/objects/AuditableFile.java)) which actually does nothing (without warning the owner of the bot).

TODO :
- [ ] Warn owner if wanted file is a directory
    - Message should tell the owner that this audit source will not be used
- [ ] Warn owner if wanted file already exists and is not writable
    - Message should tell the owner to run the bot in admin mode if he wants to use this source, and it is deactivated for now
- [ ] Warn owner if file already exists and is not a file that looks like it's for our audit
    - Message should confirm with the owner that this source is really the one he wanted
- [ ] Warn owner if file wanted does not exists but couldn't be writable
    - Message should tell the owner to run in admin mode to activate this source as we don't have the permission to write as is

**Comment by V-ed** [comment]
@StephanoMehawej, can you take care of this issue? You could actually base yourself out of our TPs from Soti, I remember that she does handle file/folder creation.

This would remove some work off my shoulders, the javadoc and unit tests are two very big issues which I don't think you'll want to dive too much into yourself x)

**Comment by StephanoMehawej** [comment]
@V-ed I am not sure of exactly what this issue is.

**Comment by V-ed** [comment]
https://github.com/Vhoyon/Discord-Bot/blob/2c5f5889a79db73756c1ea197d7620203512b103/framework/vendor/objects/AuditableFile.java#L27-L43

As you can see in this code snippet, I get the file and write to it directly. While this may work in perfect conditions, there is many cases where this method could crash and not warn the owner that a problem happened :
- The path given is not a file, but rather a folder. We cannot write data to a folder, only files. We need to handle if the path given is a folder and warn the owner that he messed up his path to an audit file.
- The file might not be writable because it's in directories protected by the system. In these cases, the bot might be able to write to the file if it was started using augmented priviledges (*`Run as administrator` in Windows, `sudo`/`root user` for Linux distros*).
- The file doesn't exist but we cannot create it. This is a variation of the second point, but is still a case we need to handle.
- The file exists and we can successfully write to it, *but* the content is not formatted correctly.
    - This case is a little harder to tackle on but I guess we could put a header in the file (the first line could be `[AUDIT]` or somethin') and verify that the header is okay.

Basically, we have to handle the cases put in the bullet points of the issue's description.

If it's still unclear, hit me up again!

**Comment by V-ed** [comment]
@StephanoMehawej After our discussion, here's something that I found that may be useful : https://stackoverflow.com/a/8630510/3784670

The interesting part is this :
> When you are creating `FileWriter` it calls constructor of `FileOutputStream` that calls a sequence of security checks *[...]*

and this :
>Here is how open() is defined:
>
>```java
>/**
> * Opens a file, with the specified name, for writing.
> * @param name name of file to be opened
> */
>private native void open(String name) throws FileNotFoundException;
>```

` throws FileNotFoundException` : As per [this stackoverflow post](https://stackoverflow.com/a/9620718/3784670) :
>It will throw a FileNotFoundException if the file doesn't exist and ***cannot be created***

`FileNotFoundException` is a subclass of `IOException`, we *our* try{}catch() uses. Therefore, you could add a catch clause before the `IOException` so that is handled as well :
```java
File auditFile = null;
try{
    auditFile = getFile();
    [...]
}
catch(FileNotFoundException e){
    // Deal with file not exists but cannot be created for security reasons
}
catch(IOException e){
    if(auditFile == null){
        log("An unexpected error occurred");
    }
    else{
        [...]
    }
}
```

I've thought about it and I really believe that we should put all the error handlings in the catch clause(s), because otherwise, we'll have multiple places where errors are handled (there would still be a need for a try...catch) and the code that actually does the action would be harder to read.

By handling in the catches, we are reading it as "you do this, and if errors happens, deal with them then".
